nilearn/glm/tests/test_second_level.py

```diff
@@ -224,28 +224,28 @@ def test_get_contrast():
 
 def test_infer_effect_maps():
     from nilearn.glm.second_level.second_level import _infer_effect_maps
-    # with InTemporaryDirectory():
```

Comment:
unrelated to the rest of this PR, this commented context manager would lead files being created in the root of the repo when running all tests in `nilearn/glm/tests/`

Reply:
OK this seems to lead to the test failing on windows

https://dev.azure.com/Parietal/Nilearn/_build/results?buildId=5559&view=logs&j=57060dbb-acbf-556b-a6e6-e748b9693eea&t=1aa43b1b-bae8-5c76-9387-adbcbcd666a7&l=17627

Reply:
reverted

Reply:
it is due to `model.masker_`, which keeps references to the loaded mask image (`mask_img_`, `_reporting_data["mask"]`), which keeps a file handle open for `mask.nii`
adding `del model` should fix the issue for now (though it relies on an implementation detail of CPython that is likely to change some day).

but anyway it is probably better to rely on pytest's `tmp_path` fixture, and that directory doesn't get removed so that error wouldn't happen.

I'm not sure why it keeps the file open in this case, as the mask image is loaded in memory during nilearn's `check_img`
we can add to the list of general cleanup tasks looking into open files held by niftiimages in nilearn objects and making sure we close them as soon as we can

Reply:
I can for sure open a draft PR to get that ball rolling.

I vaguely remember trying the tmp_path fixture approach and it broke the test.

Reply:
> I vaguely remember trying the tmp_path fixture approach and it broke the test.

it seems that the `write_fake_fmri_data_and_design` function unfortunately always writes its data in the current working directory so we need to change to the pytest temp dir, maybe that was the issue? I think something like this should work:

```python
def test_infer_effect_maps(tmp_path, monkeypatch):
    monkeypatch.chdir(tmp_path)
    from nilearn.glm.second_level.second_level import _infer_effect_maps

    shapes, rk = ((7, 8, 9, 1), (7, 8, 7, 16)), 3
    mask, fmri_data, design_matrices = write_fake_fmri_data_and_design(
        shapes,
        rk
    )
    second_level_input = pd.DataFrame({'map_name': ["a", "b"],
                                       'effects_map_path': [fmri_data[0],
                                                            "bar"]})
    assert _infer_effect_maps(second_level_input, "a") == [fmri_data[0]]
    with pytest.raises(ValueError, match="File not found: 'bar'"):
        _infer_effect_maps(second_level_input, "b")
    assert _infer_effect_maps([fmri_data[0]], None) == [fmri_data[0]]
    contrast = np.eye(rk)[1]
    second_level_input = [FirstLevelModel(mask_img=mask)] * 2
    for i, model in enumerate(second_level_input):
        model.fit(fmri_data[i],
                  design_matrices=design_matrices[i])
    assert len(_infer_effect_maps(second_level_input, contrast)) == 2
```

Reply:
It seems like there are a lot of tests that use this function and indeed they use `del` to clean up the data and it seems like usually include the model object as was suggested above. It would be much more work but wouldn't it be better if the function returned the paths without writing anything to disk and the tests handled what is written (maybe with a fixture)?

Reply:
@Remi-Gau sounds like a good idea. Then it would be easy to pass it the pytest fixture `tmp_path`

Reply:
so in order from short term to long term: (i) fix this test by using `tmp_path` and `monkeypatch.chdir(tmp_path)` (ii) add a parameter to the write_fake_fmri function to allow specifying the directory it uses (iii) see how we can avoid leaking open files after loading images